src/lsm/README.md

```diff
@@ -0,0 +1,111 @@
+# Tree
+## Compaction
+
+Tree compaction runs to the sound of music!
+
+Compacting LSM trees involves merging and moving tables into the next levels as needed.
+To avoid write amplification stalls and bound latency, compaction is done incrementally.
+
+A full compaction phase is denoted as a bar or measure, using terms from music notation.
+Each bar consists of `lsm_batch_multiple` beats or "compaction ticks" of work.
+A compaction beat is started asynchronously with `compact()` which takes a callback and
+the op that was just committed.
+
+A bar is split in half according to the "first" down beat and "middle" down beat.
+The first half of the bar compacts even levels while the latter compacts odd levels.
+Mutable table changes are sorted and compacted into the immutable table.
+The immutable table is compacted into level 0 during the odd level half of the bar.
+
+At any given point, there are at most levels/2 compactions running concurrently.
+The source level is denoted as `level_a` and the target level as `level_b`.
+The last level in the LSM tree has no target level so it is never a source level.
+
+Assuming a bar/`lsm_batch_multiple` of 4, the invariants can be described as follows:
+
+* At the end of every beat, there is space in mutable table for the next beat.
+* The manifest info for the compaction's input tables is updated at the end of the half-bar.
+* The manifest info for the compaction's output tables is updated during the compaction.
```

Comment:
What is meant by 'during the compaction'? That seems like it could describe any point during this process, but it's described as an invariant which makes me think it means some specific point in time.

Reply:
Good point, this is ambiguous here. I'll elaborate.

Reply:
You're right, this isn't exactly an invariant. And I think the previous line isn't even correct :facepalm:. I'll remove both; these were supposed to be about

> * The compactions' output tables are not visible until the compaction has finished.

anyway, which _is_ correct.